The purpose of this notebook will be the Coursera Capstone project

<h1>Choosing the Best Location<br>for a New Mexican Restaurant in the SLC, Utah Area</h1>

<img src="SLC_image.jpg" align="left">

<h2>Introduction</h2>

<p>A potential client desires to know the best location to set up a Mexican restaurant in Salt Lake City, Utah, USA. The client had operated a very successful Mexican restaurant for over ten years in Miami, Florida, USA. Operating in the food business has been in his family for several generations. The client relocated to Utah because many members of his family have relocated to Salt Lake City. However, he is very unfamiliar with the demographics and competition for mexican food in the greater Salt Lake City area.</p>
<p>The client wishes to learn which location in the city has the least competition for the greatest population. The client already knows that Mexican food is the third most popular type of cuisine in Utah, which is 50% higher than the national avereage<sup><a href="#fn1" id="ref1">1</a></sup>. This will be a family venue, so the potential location preferrably needs to be closer to an urban area.</p>

<h2>Data</h2>
<p>The Foursquare location data set will be used to analyze food venue locations and popularity. Population by zipcode data was gathered from the US Census Bureau<sup><a href="#fn2" id="ref2">2</a></sup>. Finding the best location for the Mexican restaurant will be discovered by analyzing these two data sets and comparing the most viable location (zipcode), which will be located farthest from its competition with the highest population.

<h3>Analysis</h3>

In [1]:
#Using Pandas libraries to read in zipcode/income data
import pandas as pd
#importing zip codes by city in Utah
city_zip_df = pd.read_csv("Utah_city_by_zipcode.csv")
city_zip_df.head()

Zip Code            City    County
0     84001        Altamont  Duchesne
1     84002         Altonah  Duchesne
2     84003   American Fork      Utah
3     84004          Alpine      Utah
4     84005  Eagle Mountain      Utah

In [2]:
#importing population by zipcode in Utah
pop_zip_df = pd.read_csv("Utah_population_by_zipcode.csv")
pop_zip_df.head()

Zip Code  Population
0     84001         757
1     84002         381
2     84003       46939
3     84004       10256
4     84005       27533

In [3]:
#merging both dataframes on zipcode
city_zip_pop = pd.merge(city_zip_df, pop_zip_df, on='Zip Code')
city_zip_pop.head()

Zip Code            City    County  Population
0     84001        Altamont  Duchesne         757
1     84002         Altonah  Duchesne         381
2     84003   American Fork      Utah       46939
3     84004          Alpine      Utah       10256
4     84005  Eagle Mountain      Utah       27533

In [4]:
#finally, we only need zipcodes found in Salt Lake County
sl_county = city_zip_pop[city_zip_pop['County'] == "Salt Lake"]
sl_county.head()

Zip Code            City     County  Population
5      84006  Bingham Canyon  Salt Lake         579
13     84020          Draper  Salt Lake       46111
33     84044           Magna  Salt Lake       28257
36     84047         Midvale  Salt Lake       33343
52     84065        Riverton  Salt Lake       41838

In [5]:
#resetting index values
sl_county.reset_index(drop=True, inplace=True)

In [6]:
sl_county.head()

Zip Code            City     County  Population
0     84006  Bingham Canyon  Salt Lake         579
1     84020          Draper  Salt Lake       46111
2     84044           Magna  Salt Lake       28257
3     84047         Midvale  Salt Lake       33343
4     84065        Riverton  Salt Lake       41838

In [7]:
#read in latitude-longitude coordinates database for Utah
longlat = pd.read_csv("Utah_lat_long.csv")
longlat.head()

Zip Code   Latitude   Longitude
0     84530  38.236886 -109.154450
1     84305  41.938629 -112.043300
2     84307  41.535714 -112.303730
3     84133  40.769046 -111.889319
4     13503  43.101869  -75.231158

In [8]:
#merge with sl_county dataframe
sl_county_new = pd.merge(sl_county, longlat, on='Zip Code')
sl_county_new.head()

Zip Code            City     County  Population   Latitude  Longitude
0     84006  Bingham Canyon  Salt Lake         579  40.563896 -112.12810
1     84020          Draper  Salt Lake       46111  40.514843 -111.87294
2     84044           Magna  Salt Lake       28257  40.700162 -112.08318
3     84047         Midvale  Salt Lake       33343  40.614997 -111.88704
4     84065        Riverton  Salt Lake       41838  40.503746 -111.96327

In [9]:
#checking shape
sl_county_new.shape

(35, 6)

In [10]:
#renaming "Zip Code" column header to "Zipcode"
sl_county_new.rename(columns = {'Zip Code':'Zipcode'}, inplace=True)
sl_county_new

Zipcode            City     County  Population   Latitude   Longitude
0     84006  Bingham Canyon  Salt Lake         579  40.563896 -112.128100
1     84020          Draper  Salt Lake       46111  40.514843 -111.872940
2     84044           Magna  Salt Lake       28257  40.700162 -112.083180
3     84047         Midvale  Salt Lake       33343  40.614997 -111.887040
4     84065        Riverton  Salt Lake       41838  40.503746 -111.963270
5     84070           Sandy  Salt Lake       27736  40.581595 -111.888210
6     84084     West Jordan  Salt Lake       37655  40.627668 -111.975660
7     84088     West Jordan  Salt Lake       37742  40.594222 -111.978890
8     84092           Sandy  Salt Lake       29498  40.558748 -111.809020
9     84093           Sandy  Salt Lake       23294  40.594948 -111.834480
10    84094           Sandy  Salt Lake       28465  40.569598 -111.862170
11    84095    South Jordan  Salt Lake       65890  40.561598 -111.960880
12    84101  Salt Lake City  Salt Lake        4653  40.754746 -111.898750
13    84102  Salt Lake City  Salt Lake       17859  40.759246 -111.863710
14    84103  Salt Lake City  Salt Lake       22202  40.777746 -111.874910
15    84104  Salt Lake City  Salt Lake       25260  40.750628 -111.940770
16    84105  Salt Lake City  Salt Lake       22507  40.737132 -111.860440
17    84106  Salt Lake City  Salt Lake       36160  40.706246 -111.856570
18    84107  Salt Lake City  Salt Lake       34302  40.659997 -111.883940
19    84108  Salt Lake City  Salt Lake       22071  40.744646 -111.811230
20    84109  Salt Lake City  Salt Lake       24195  40.705747 -111.799730
21    84111  Salt Lake City  Salt Lake       11329  40.754700 -111.883610
22    84112  Salt Lake City  Salt Lake        1916  40.767840 -111.838050
23    84113  Salt Lake City  Salt Lake         210  40.763322 -111.833640
24    84115  Salt Lake City  Salt Lake       26832  40.718113 -111.889380
25    84116  Salt Lake City  Salt Lake       35841  40.790309 -111.940180
26    84117  Salt Lake City  Salt Lake       25057  40.662797 -111.837970
27    84118  Salt Lake City  Salt Lake       71185  40.650496 -111.995910
28    84119  Salt Lake City  Salt Lake       55580  40.696596 -111.952390
29    84120  Salt Lake City  Salt Lake       51270  40.688246 -111.998990
30    84121  Salt Lake City  Salt Lake       40665  40.625621 -111.815800
31    84123  Salt Lake City  Salt Lake       39122  40.659514 -111.922260
32    84124  Salt Lake City  Salt Lake       22180  40.680064 -111.822200
33    84128  Salt Lake City  Salt Lake       26574  40.694746 -112.040040
34    84144  Salt Lake City  Salt Lake         175  40.767946 -111.890819

In [11]:
#importing necessary libraries to begin exploring data
import numpy as np
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)
import json # library to handle JSON files
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values
import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe
# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors
# import k-means from clustering stage
from sklearn.cluster import KMeans
import folium # map rendering library
print('Libraries imported.')

Libraries imported.


In [12]:
#using geopy to get the latitude and longitude values of Salt Lake City area

address = 'Salt Lake City, UT'

geolocator = Nominatim(user_agent="SLC_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Salt Lake City are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Salt Lake City are 40.7670126, -111.8904308.


In [13]:
#creating map of Salt Lake County with the zipcodes areas superimposed on top

# create map of Salt Lake City area using latitude and longitude values
map_SLC = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, zipcode, city, pop in zip(sl_county_new['Latitude'], sl_county_new['Longitude'], sl_county_new['Zipcode'], sl_county_new['City'], sl_county_new['Population']):
    label = 'zip:{}, city:{}, pop:{}'.format(zipcode, city, pop)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_SLC)  
    
map_SLC

In [14]:
#Exploring the data using Foursquare.
CLIENT_ID = 'T5NN5R3X5MYCHLPUSROQ2G2C113441HZWFLEWS51UZYL5RQE' # Foursquare ID
CLIENT_SECRET = '0XBARW5IONQUPDPGFSFA31IXZBT10JIR2VMDQERLHM02GSGZ' # Foursquare Secret
VERSION = '20180604'

In [15]:
#function to get nearby Mexican food venues by zipcode

def getNearbyVenues(names, latitudes, longitudes, radius=500):
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}&categoryId=4bf58dd8d48988d1c1941735'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'],
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Zipcode',         
                  'Zipcode Latitude', 
                  'Zipcode Longitude', 
                  'Venue',     
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [16]:
LIMIT = 100 # limit of number of venues returned by Foursquare API
radius = 10000 # This is a 6.5 mile radius which is sufficient to cover all of SL County

In [17]:
#creating code to run the above function on each Zipcode and create a new dataframe called slc_venues

slc_venues = getNearbyVenues(names=sl_county_new['Zipcode'],
                                   latitudes=sl_county_new['Latitude'],
                                   longitudes=sl_county_new['Longitude']
                                  )

84006
84020
84044
84047
84065
84070
84084
84088
84092
84093
84094
84095
84101
84102
84103
84104
84105
84106
84107
84108
84109
84111
84112
84113
84115
84116
84117
84118
84119
84120
84121
84123
84124
84128
84144


In [18]:
slc_venues

Zipcode  Zipcode Latitude  Zipcode Longitude  \
0     84044         40.700162        -112.083180   
1     84047         40.614997        -111.887040   
2     84070         40.581595        -111.888210   
3     84070         40.581595        -111.888210   
4     84070         40.581595        -111.888210   
5     84093         40.594948        -111.834480   
6     84093         40.594948        -111.834480   
7     84095         40.561598        -111.960880   
8     84101         40.754746        -111.898750   
9     84102         40.759246        -111.863710   
10    84102         40.759246        -111.863710   
11    84106         40.706246        -111.856570   
12    84107         40.659997        -111.883940   
13    84111         40.754700        -111.883610   
14    84112         40.767840        -111.838050   
15    84116         40.790309        -111.940180   
16    84117         40.662797        -111.837970   
17    84117         40.662797        -111.837970   
18    84117         40.662797        -111.837970   
19    84119         40.696596        -111.952390   
20    84119         40.696596        -111.952390   
21    84119         40.696596        -111.952390   
22    84119         40.696596        -111.952390   
23    84144         40.767946        -111.890819   
24    84144         40.767946        -111.890819   
25    84144         40.767946        -111.890819   
26    84144         40.767946        -111.890819   
27    84144         40.767946        -111.890819   
28    84144         40.767946        -111.890819   
29    84144         40.767946        -111.890819   
30    84144         40.767946        -111.890819   

                           Venue  Venue Latitude  Venue Longitude  \
0                    El Habanero       40.696706      -112.086080   
1                        Morelia       40.618763      -111.885268   
2     Cantina Southwestern Grill       40.581923      -111.888617   
3              The Mayan Express       40.581291      -111.890203   
4       Chile Verde at Rio Tinto       40.583446      -111.893128   
5           Altamar Mexican Food       40.593829      -111.832281   
6        La Hacienda de Don Juan       40.594504      -111.831856   
7             Paul Bunyans Cabin       40.562561      -111.966408   
8         Alberto's Mexican Food       40.758170      -111.899092   
9                       Del Taco       40.760877      -111.867938   
10                 Lana Del Taco       40.760929      -111.867907   
11                      Del Taco       40.705765      -111.853336   
12        Chipotle Mexican Grill       40.659763      -111.887871   
13                     Taco Taco       40.758402      -111.885026   
14           El Sillero by Fiana       40.768555      -111.842585   
15        Hacienda Mexican Grill       40.785992      -111.940742   
16                  Rancherito's       40.666745      -111.839032   
17                 Real Taqueria       40.666030      -111.838860   
18                     Taco Bell       40.666463      -111.840510   
19        Cafe Rio Mexican Grill       40.695090      -111.957412   
20                Cafe Silvestre       40.696107      -111.948803   
21    Crazy Lizard Mexican Grill       40.692919      -111.954226   
22           Taco Daniel Express       40.693092      -111.955276   
23           Taste of Red Iguana       40.768264      -111.889502   
24                   Blue Iguana       40.765804      -111.893271   
25                    Costa Vida       40.768798      -111.889360   
26                   Taqueria 27       40.765168      -111.886339   
27  City Creek Center Food Court       40.768258      -111.889498   
28                   Cancun Cafe       40.765104      -111.887407   
29         Maize Homestyle Tacos       40.765990      -111.889687   
30                    Real Tacos       40.765009      -111.888127   

        Venue Category  
0   Mexican Restaurant  
1   Mexican Restaurant  
2   Mexican Restaurant  
3   Mexican Restaurant  
4    

In [19]:
#checking to see how many venues were returned for each Zipcode

slc_venues.groupby('Zipcode').count()

Zipcode Latitude  Zipcode Longitude  Venue  Venue Latitude  \
Zipcode                                                               
84044                   1                  1      1               1   
84047                   1                  1      1               1   
84070                   3                  3      3               3   
84093                   2                  2      2               2   
84095                   1                  1      1               1   
84101                   1                  1      1               1   
84102                   2                  2      2               2   
84106                   1                  1      1               1   
84107                   1                  1      1               1   
84111                   1                  1      1               1   
84112                   1                  1      1               1   
84116                   1                  1      1               1   
84117                   3                  3      3               3   
84119                   4                  4      4               4   
84144                   8                  8      8               8   

         Venue Longitude  Venue Category  
Zipcode                                   
84044                  1               1  
84047                  1               1  
84070                  3               3  
84093                  2               2  
84095                  1               1  
84101                  1               1  
84102                  2               2  
84106                  1               1  
84107                  1               1  
84111                  1               1  
84112                  1               1  
84116                  1               1  
84117                  3               3  
84119                  4               4  
84144                  8               8

In [20]:
# creat new count dataframe by zipcode
venue_count = slc_venues.groupby('Zipcode').count()
venue_count.head()

Zipcode Latitude  Zipcode Longitude  Venue  Venue Latitude  \
Zipcode                                                               
84044                   1                  1      1               1   
84047                   1                  1      1               1   
84070                   3                  3      3               3   
84093                   2                  2      2               2   
84095                   1                  1      1               1   

         Venue Longitude  Venue Category  
Zipcode                                   
84044                  1               1  
84047                  1               1  
84070                  3               3  
84093                  2               2  
84095                  1               1

In [21]:
# drop all the unneeded columns
venue_count.drop(['Zipcode Latitude','Zipcode Longitude','Venue Latitude','Venue Longitude','Venue Category'], axis=1, inplace=True)
venue_count.head()

Venue
Zipcode       
84044        1
84047        1
84070        3
84093        2
84095        1

In [22]:
# rename "Venue" column header to "Count"
venue_count.rename(columns = {'Venue':'Count'}, inplace=True)
venue_count

Count
Zipcode       
84044        1
84047        1
84070        3
84093        2
84095        1
84101        1
84102        2
84106        1
84107        1
84111        1
84112        1
84116        1
84117        3
84119        4
84144        8

In [23]:
# Now, merging the venue_count dataframe with sl_county_new
sl_county_merged = pd.merge(sl_county_new, venue_count, on='Zipcode')
sl_county_merged

Zipcode            City     County  Population   Latitude   Longitude  \
0     84044           Magna  Salt Lake       28257  40.700162 -112.083180   
1     84047         Midvale  Salt Lake       33343  40.614997 -111.887040   
2     84070           Sandy  Salt Lake       27736  40.581595 -111.888210   
3     84093           Sandy  Salt Lake       23294  40.594948 -111.834480   
4     84095    South Jordan  Salt Lake       65890  40.561598 -111.960880   
5     84101  Salt Lake City  Salt Lake        4653  40.754746 -111.898750   
6     84102  Salt Lake City  Salt Lake       17859  40.759246 -111.863710   
7     84106  Salt Lake City  Salt Lake       36160  40.706246 -111.856570   
8     84107  Salt Lake City  Salt Lake       34302  40.659997 -111.883940   
9     84111  Salt Lake City  Salt Lake       11329  40.754700 -111.883610   
10    84112  Salt Lake City  Salt Lake        1916  40.767840 -111.838050   
11    84116  Salt Lake City  Salt Lake       35841  40.790309 -111.940180   
12    84117  Salt Lake City  Salt Lake       25057  40.662797 -111.837970   
13    84119  Salt Lake City  Salt Lake       55580  40.696596 -111.952390   
14    84144  Salt Lake City  Salt Lake         175  40.767946 -111.890819   

    Count  
0       1  
1       1  
2       3  
3       2  
4       1  
5       1  
6       2  
7       1  
8       1  
9       1  
10      1  
11      1  
12      3  
13      4  
14      8

In [24]:
# finally, adding a new column, "Count/Population," which will be the result of dividing the Count (of Mexican Restaurants)
# by the Population of each Zipcode
sl_county_merged['Count/Population'] = sl_county_merged['Count']/sl_county_merged['Population']
sl_county_merged

Zipcode            City     County  Population   Latitude   Longitude  \
0     84044           Magna  Salt Lake       28257  40.700162 -112.083180   
1     84047         Midvale  Salt Lake       33343  40.614997 -111.887040   
2     84070           Sandy  Salt Lake       27736  40.581595 -111.888210   
3     84093           Sandy  Salt Lake       23294  40.594948 -111.834480   
4     84095    South Jordan  Salt Lake       65890  40.561598 -111.960880   
5     84101  Salt Lake City  Salt Lake        4653  40.754746 -111.898750   
6     84102  Salt Lake City  Salt Lake       17859  40.759246 -111.863710   
7     84106  Salt Lake City  Salt Lake       36160  40.706246 -111.856570   
8     84107  Salt Lake City  Salt Lake       34302  40.659997 -111.883940   
9     84111  Salt Lake City  Salt Lake       11329  40.754700 -111.883610   
10    84112  Salt Lake City  Salt Lake        1916  40.767840 -111.838050   
11    84116  Salt Lake City  Salt Lake       35841  40.790309 -111.940180   
12    84117  Salt Lake City  Salt Lake       25057  40.662797 -111.837970   
13    84119  Salt Lake City  Salt Lake       55580  40.696596 -111.952390   
14    84144  Salt Lake City  Salt Lake         175  40.767946 -111.890819   

    Count  Count/Population  
0       1          0.000035  
1       1          0.000030  
2       3          0.000108  
3       2          0.000086  
4       1          0.000015  
5       1          0.000215  
6       2          0.000112  
7       1          0.000028  
8       1          0.000029  
9       1          0.000088  
10      1          0.000522  
11      1          0.000028  
12      3          0.000120  
13      4          0.000072  
14      8          0.045714

In [25]:
# sorting values by "Count/Population" column
sl_county_merged.sort_values('Count/Population')

Zipcode            City     County  Population   Latitude   Longitude  \
4     84095    South Jordan  Salt Lake       65890  40.561598 -111.960880   
7     84106  Salt Lake City  Salt Lake       36160  40.706246 -111.856570   
11    84116  Salt Lake City  Salt Lake       35841  40.790309 -111.940180   
8     84107  Salt Lake City  Salt Lake       34302  40.659997 -111.883940   
1     84047         Midvale  Salt Lake       33343  40.614997 -111.887040   
0     84044           Magna  Salt Lake       28257  40.700162 -112.083180   
13    84119  Salt Lake City  Salt Lake       55580  40.696596 -111.952390   
3     84093           Sandy  Salt Lake       23294  40.594948 -111.834480   
9     84111  Salt Lake City  Salt Lake       11329  40.754700 -111.883610   
2     84070           Sandy  Salt Lake       27736  40.581595 -111.888210   
6     84102  Salt Lake City  Salt Lake       17859  40.759246 -111.863710   
12    84117  Salt Lake City  Salt Lake       25057  40.662797 -111.837970   
5     84101  Salt Lake City  Salt Lake        4653  40.754746 -111.898750   
10    84112  Salt Lake City  Salt Lake        1916  40.767840 -111.838050   
14    84144  Salt Lake City  Salt Lake         175  40.767946 -111.890819   

    Count  Count/Population  
4       1          0.000015  
7       1          0.000028  
11      1          0.000028  
8       1          0.000029  
1       1          0.000030  
0       1          0.000035  
13      4          0.000072  
3       2          0.000086  
9       1          0.000088  
2       3          0.000108  
6       2          0.000112  
12      3          0.000120  
5       1          0.000215  
10      1          0.000522  
14      8          0.045714

In [26]:
# further analysis of South Jordan; looking for the most popular venue in order to
# position the restaurant where the most people gather!
# Querying Foursquare again, but for South Jordan lat/lng

In [27]:
# setting lat/lng variables for South Jordan and other variables

lat = 40.561598
lng = -111.960880
radius = 2500
LIMIT = 50

SJ_venues = []

# creating URL to query Foursquare and order returned venues by popularity

url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}&sortByPopularity=1'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    lat, 
    lng, 
    radius, 
    LIMIT)
            
# make the GET request
results = requests.get(url).json()["response"]['groups'][0]['items']
        
# return only relevant information for each nearby venue
SJ_venues.append([( 
    v['venue']['name'],
    v['venue']['location']['lat'], 
    v['venue']['location']['lng'],  
    v['venue']['categories'][0]['name']) for v in results])

nearby_SJ_venues = pd.DataFrame([item for SJ_venues in SJ_venues for item in SJ_venues])
nearby_SJ_venues.columns = [ 
                  'Venue',     
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
nearby_SJ_venues.head()

Venue  Venue Latitude  Venue Longitude        Venue Category
0           Costco       40.559808      -111.975857       Warehouse Store
1          Walmart       40.563091      -111.975016         Big Box Store
2     Vasa Fitness       40.560443      -111.979258  Gym / Fitness Center
3  Harmons Grocery       40.543447      -111.977755         Grocery Store
4  Harmons Grocery       40.560470      -111.936935         Grocery Store

In [ ]:
# Putting the restaurant next to Costco or Walmart would be a big score! Let's look at the map

In [32]:
#creating map of South Jordan area with the top venues superimposed on top. First setting
#variables for latitude and longitude for South Jordan

latitude = 40.561598
longitude = -111.960880

# create map of Salt Lake City area using latitude and longitude values
map_SJ = folium.Map(location=[latitude, longitude], zoom_start=15)

# add markers to map
for lat, lng, venue in zip(nearby_SJ_venues['Venue Latitude'], nearby_SJ_venues['Venue Longitude'], nearby_SJ_venues['Venue']):
    label = 'Venue:{}'.format(venue)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_SJ)  
    
map_SJ

In [ ]:
# Wow, look at all the popular venues clustered by Costco and Walmart! I wonder why?!

<h2>Summary Observations and Recommendation</h2>
<p>By inspection, the Zipcode that has the lowest number of Mexican Restaurants per person is 84095 (South Jordan, UT) at .000015 Restaurants/Person! This fits very well with common sense because South Jordan is the fastest growing city in Salt Lake County with a mean income of $119,416<sup><a href="#fn3" id="ref3">3</a></sup>.</p><br>
<p>Costco and Walmart are big magnets to the population. Therefore, placing a new Mexican Restaurant in South Jordan near Costco and Walmart looks very promising and will be recommended to the client.</p>

<hr>
<P><ul>End Notes</ul><br>
<sup id="fn1">1. https://fox13now.com/2015/01/16/utahs-favorite-cuisine-is-what-most-disproportionately-popular-cuisine-by-state/<a href="#ref1" title="Jump back to footnote 1 in the text.">↩</a></sup><br>
<sup id="fn2">2. https://www.utah-demographics.com/cities_by_population<a href="#ref2" title="Jump back to footnote 2 in the text.">↩</a></sup><br>
<sup id="fn3">3. https://www.incomebyzipcode.com/utah<a href="#ref3" title="Jump back to footnote 3 in the text.">↩</a></sup>